### PR TITLE
Add sorting by key to Tables json.pretty()

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Implement Tables.sorted_pairs with an user-specified comparator
Add lexicographical ordering to json.pretty()

### CODE REVIEW

File: lib/pure/collections/tables.nim [hunk 1]
```diff
@@ -240,6 +240,26 @@ iterator pairs*[A, B](t: Table[A, B]): (A, B) =
   for h in 0..high(t.data):
     if isFilled(t.data[h].hcode): yield (t.data[h].key, t.data[h].val)
 
+import algorithm
+
+iterator sorted_pairs*[A, B](t: Table[A, B],
```
Comment (Araq) [inline lib/pure/collections/tables.nim]: NEP1 says it needs to be sortedPairs everywhere.

File: lib/pure/collections/tables.nim [hunk 2]
```diff
@@ -240,6 +240,26 @@ iterator pairs*[A, B](t: Table[A, B]): (A, B) =
...
+  order = SortOrder.Ascending): (A, B) =
+  ## iterates over any (key, value) pair in the table `t` in sorted order
+  ## based on the comparator `cmp`
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   for key, val in sorted_pairs(t, system.cmp[string]):
+  ##     ...
+  ##
+  var keys: seq[A] = @[]
```
Comment (Araq) [inline lib/pure/collections/tables.nim]: you can use 'newSeqOfCap(t.len)' here

File: lib/pure/json.nim [hunk 3]
```diff
@@ -964,16 +964,28 @@ proc escapeJson*(s: string): string =
...
   of JObject:
     if currIndent != 0 and not lstArr: result.nl(ml)
     result.indent(currIndent) # Indentation
     if node.fields.len > 0:
       result.add("{")
       result.nl(ml) # New line
+      var get_pairs = (block:
+        iterator it: (string, JsonNode) {.closure.} =
+          if sort_keys:
+            for key, val in sorted_pairs(node.fields, system.cmp[string]):
```
Comment (Araq) [inline lib/pure/json.nim]: 'cmp' without the string specializiation produces better code. (And I hope it compiles)

**Comment by Araq** [comment]
Nice, but I think the Nimble problem needs to be solved by Json using OrderedTables instead of Tables.

File: lib/pure/json.nim [hunk 4]
```diff
@@ -964,16 +964,28 @@ proc escapeJson*(s: string): string =
...
+              lstArr = false, currIndent = 0, sort_keys = false) =
   case node.kind
   of JObject:
     if currIndent != 0 and not lstArr: result.nl(ml)
     result.indent(currIndent) # Indentation
     if node.fields.len > 0:
       result.add("{")
       result.nl(ml) # New line
+      var get_pairs = (block:
+        iterator it: (string, JsonNode) {.closure.} =
```
Comment (yglukhov) [inline lib/pure/json.nim]: Closure iterators are not supported by js backend.

**Comment by dom96** [comment]
Are we still going to accept this despite the move to OrderedTables?

**Comment by Araq** [comment]
> Are we still going to accept this despite the move to OrderedTables?

I like it but yglukhov's remark is a biggie. So it cannot be merged until the JS codegen supports first class iterators.

**Comment by yglukhov** [comment]
Actually this pr can (and imho should) easily be rewritten to not use closures at least to reduce gc pressure.

**Comment by dom96** [comment]
Should we close this now that the json module uses an ordered table?

**Comment by FedericoCeratto** [comment]
@dom96 being able to generate sorted JSON is still useful (especially when JSON files are compared by VCSes) without having to reorder tables

**Comment by dom96** [comment]
Okay, but then this PR needs to be simplified.

File: lib/pure/collections/tables.nim [hunk 5]
```diff
@@ -240,6 +240,26 @@ iterator pairs*[A, B](t: Table[A, B]): (A, B) =
   for h in 0..high(t.data):
     if isFilled(t.data[h].hcode): yield (t.data[h].key, t.data[h].val)
 
+import algorithm
+
+iterator sortedPairs*[A, B](t: Table[A, B],
+  cmp: proc (x, y: A): int {.closure.};
```
Comment (yglukhov) [inline lib/pure/collections/tables.nim]: This doesnt have to be closure, does it?

File: lib/pure/collections/tables.nim [hunk 6]
```diff
@@ -240,6 +240,26 @@ iterator pairs*[A, B](t: Table[A, B]): (A, B) =
...
+  order = SortOrder.Ascending): (A, B) =
+  ## iterates over any (key, value) pair in the table `t` in sorted order
+  ## based on the comparator `cmp`
+  ##
+  ## .. code-block:: nim
+  ##
+  ##   for key, val in sortedPairs(t, system.cmp[string]):
+  ##     ...
+  ##
+  var keys: seq[A] = @[]
```
Comment (yglukhov) [inline lib/pure/collections/tables.nim]: Use `toSeq` or at least preallocate the seq with correct size

File: lib/pure/collections/tables.nim (hunk 5 above)
Comment (yglukhov) [inline lib/pure/collections/tables.nim]: Tbh, i'm not even sure if its a good idea to expose this as a public proc. Such api can easily be misused in cases where OrderedTable would be more appropriate.

**Comment by Araq** [comment]
No updates, closing. The feature seems overkill to me.

**Pull request closed without merging**